src/lib/alloc.c

```diff
@@ -945,7 +942,8 @@ static void *_balloc_unlocked(uint32_t flags, uint32_t caps, size_t bytes,
 		/* Continue from the next heap */
 	}
 
-	return ptr;
+	return (flags & SOF_MEM_FLAG_COHERENT) && (CONFIG_CORE_COUNT > 1) ?
+		cache_to_uncache(ptr) : uncache_to_cache(ptr);
```

Comment:
Does this mean, that when building for a single core, we disable all coherency management and force all memory to be cached (except DMA addresses)? I guess this can be done, but isn't it too confusing and risky? We would be relying on the correct use of the `SOF_MEM_FLAG_COHERENT` *flag* to indicate "coherent on multi-core, not used for DMA" as opposed to the `SOF_MEM_CAPS_DMA` *capability* to indicate "uncached / coherent for DMA." @lgirdwood do we really need separate `flags` and `caps` for `rmalloc()` and friends? Maybe we could get rid of one of them and streamline this a bit?

Reply:
yes, it's tricky at the moment as we don't have @ranj063 's coherent API yet, the "(CONFIG_CORE_COUNT > 1)" was added in my latest version to keep the behavior and avoid introducing regressions on APL platforms, via keeping intact of using always cached address. Otherwise, we will see many failures on APL platforms in CI.

Once @ranj063 's coherent APIs work well, we should remove this "(CONFIG_CORE_COUNT > 1)" check and return type with checking the flags only.